### ISSUE TITLE
Add --cap-add and --cap-drop support

### ISSUE THREAD

**Issue by darthlukan** [containers/toolbox#1042]
**Is your feature request related to a problem? Please describe.**
Unable to write to `/dev/mmcblkX` devices from within a toolbox container. This is necessary for various development environments where one may, such as in my case, be automating the provisioning of SDCARDs with custom images or firmware.

**Describe the solution you'd like**
I would like to be able to do something like: `toolbox enter firmware-dev --cap-add CAP_SYS_RAWIO` and `toolbox enter firmware-dev --cap-drop CAP_SYS_RAWIO` or do the same with any other Linux capability.

**Describe alternatives you've considered**
The alternative is to effectively reproduce what `toolbox` is doing with `podman` under the hood via a script wrapping `podman` directly. This is less than ideal just for the sheer amount of typing involved :stuck_out_tongue:

**Comment by HarryMichal** [comment]
Hi @darthlukan! Thanks for filing this feature request!

We're generally opposed to the idea of adding additional knobs and switches to Toolbx simply for the sake of trying to keep its complexity at bay. Your request is making me wonder, though. Consider you're running on a "classic" distro (e.g., Fedora Workstation), how does the user get the `CAP_SYS_RAWIO` capability? Is that something a generic user account has? Or is this an action one would reach for `sudo`/`pkexec`? The reason why I'm asking is that we can not grant any capabilities in the container that its creator can assume.

I'm know very little when it comes to "capabilities", so any insight into the matter is more than welcome.

**Comment by darthlukan** [comment]
Hi! I'm not 100% certain myself, but I do believe that a user can only consume capabilities that they have access to, so a user may have to use `sudo` for some things if their user is not part of a group that allows say, writing to IO devices.

For example, if a user is in the `console` group, they may be able to write to `/dev/USBtty0` or if in the `disk` group, able to write to `/dev/sda`. Normally though, users would not be in these groups, they'd use `su | sudo` for the requisite write permissions. In this case, the user being `su | sudo`'d into would have to execute the toolbox/container. I think this is a sane compromise. 

Detecting executing user ownership before processing the `CAPABILITY` flag may be too much as a first pass, but perhaps noting that "You must use `sudo` to add the desired capability" is a much easier one.

**Comment by HarryMichal** [comment]
> In this case, the user being su | sudo'd into would have to execute the toolbox/container.

Yes, this is something I was aiming towards. It can be quite inconvenient having to discern between "rootfull" and "rootless" containers but that's the cost of the ecosystem.

> ...but perhaps noting that "You must use sudo to add the desired capability" is a much easier one.

We could clarify this in the manual.

One thing that might be worth checking is whether your account's membership in a particular group is translated into a toolbx. If not, then we probably have a bug on our hands.

**Comment by akdev1l** [comment]
on my Fedora 36 system a non-root user has no capabilities:

```
[akdev@canzuk ~]$ awk /Cap/ /proc/$$/status
CapInh:	0000000000000000
CapPrm:	0000000000000000
CapEff:	0000000000000000
CapBnd:	000001ffffffffff
CapAmb:	0000000000000000
```

I'll have to double check if this is standard configuration but if that's the case then there's nothing we could do to acquire the capabilities as non-root.

the `CapPrm` indicates the permitted set of capabilities that we can add `CapEff` which is the effective capabilities of the process (in this example it is my shell). They are both `0x0` so we can't add capabilities new ones.

I'm not entirely sure how the capabilities subsystem interacts with containers - as far as I know it is not namespaced... 

from a non-root toolbox container I see that even as "root" I cannot get pass `CAP_NET_RAW`:

```
⬢[akdev@test4 ~]$ awk /Cap/ /proc/$$/status
CapInh:	0000000000000000
CapPrm:	0000000000000000
CapEff:	0000000000000000
CapBnd:	000001ffffffffff
CapAmb:	0000000000000000
⬢[akdev@test4 ~]$ ping -vc1 google.com
ping: socket: Operation not permitted
ping: socket: Operation not permitted
```

even if a root process inside the container is reported that it has the capabilities it really doesn't:

```
⬢[akdev@test4 ~]$ sudo sleep 100 &
[1] 507731
⬢[akdev@test4 ~]$ awk /Cap/ /proc/507731/status
CapInh:	0000000000000000
CapPrm:	000001ffffffffff
CapEff:	000001ffffffffff
CapBnd:	000001ffffffffff
CapAmb:	0000000000000000
```

I don't think we'd be able to support this feature on rootless containers - and I think it should work already on rootful containers (those are `--privileged` already if I'm not mistaken)

**Comment by heyakyra** [comment]
Ah, this must be why I hit this: https://github.com/containers/toolbox/pull/483#issuecomment-1224565572 !

**Comment by debarshiray** [comment]
> on my Fedora 36 system a non-root user has no capabilities:
> ...
> I'll have to double check if this is standard configuration but if
> that's the case then there's nothing we could do to acquire the
> capabilities as non-root.

Yes, that's the standard configuration.

> from a non-root toolbox container I see that even as "root" I
> cannot get pass CAP_NET_RAW:

Yes, that's because `root` inside a rootless container isn't actually `root` or UID 0 on the host.

However, you should [be able](https://github.com/systemd/systemd/commit/90ce7627dfe824ff6e7c0ca5f96350fbcfec7118) to use `ping(8)` via [ICMP sockets](https://lwn.net/Articles/422330/) on Fedora.

**Comment by KilianHanich** [comment]
> One thing that might be worth checking is whether your account's membership in a particular group is translated into a toolbx. If not, then we probably have a bug on our hands.

Seems like there is:

![image](https://github.com/containers/toolbox/assets/69722661/4a500800-35b1-4330-86ce-2aa18fcf1a06)

But while this is an arbitrary group (which you can guess by the name I guess), it should  keep the id if possible.

Files are handled by translating all these to "nobody" for both, groups and users which I am not sure should be done. (And yes, that means everything under /dev in my toolboxes is owned by "nobody" and has the "nobody" group.)

**Comment by darthlukan** [comment]
Wow! It has been a while since this conversation started. I've since moved-on from the project that was the impetus for the feature request, and my work-around at the time was to just use `sudo || su` via `become` as needed within my Ansible playbooks to do the requisite work in an elevated privilege container.

I think, given the discussion content, age, and uncommon use-case which is beyond the intended scope of Toolbox, this feature request can be closed.
